src/WebJobs.Extensions.DurableTask/Listener/TaskOrchestrationShim.cs

```diff
@@ -178,6 +131,65 @@ public override async Task<string> Execute(OrchestrationContext innerContext, st
             return serializedOutput;
         }
 
+        private async Task<object> InvokeFunctionAsync(RegisteredFunctionInfo orchestratorInfo)
+        {
+            try
+            {
+                Task invokeTask = this.functionInvocationCallback();
+                if (invokeTask is Task<object> resultTask)
+                {
+                    // Orchestrator threads cannot perform async I/O, so block on such out-of-proc threads.
+                    // Possible performance implications; may need revisiting.
+                    return orchestratorInfo.IsOutOfProc ? resultTask.Result : await resultTask;
+                }
+                else
+                {
+                    throw new InvalidOperationException("The WebJobs runtime returned a invocation task that does not support return values!");
+                }
+            }
+            catch (Exception e)
+            {
+                if (OutOfProcExceptionHelpers.TryExtractOutOfProcStateJson(e.InnerException, out string returnValue)
+                    && !string.IsNullOrEmpty(returnValue))
+                {
+                    return returnValue;

```

Comment:
If we get here, does it mean the orchestrator function failed with an unhandled error? If so, don't we still need to trace `FunctionFailed` and send the `OrchestratorFailed` lifecycle event?

Reply:
Hmm probably. At least part of this is handled in `HandleOutOfProcExecution()` in the below code:

```csharp
if (!string.IsNullOrEmpty(execution.Error))
{
    var orchestrationException = new OrchestrationFailureException(
    $"Orchestrator function '{this.context.Name}' failed: {execution.Error}");
    this.context.OrchestrationException = ExceptionDispatchInfo.Capture(orchestrationException);
    throw orchestrationException;
}
```

That being said, the above code block is missing some logging and the lifecycle event. I will factor out the rest of this `catch` block into a method that can be called from within `HandleOutOfProcExecution()`.

Reply:
I have completely refactored how I handle this, due to needing the `Exception` thrown here to be available in `HandleOutOfProcExecution`. Would love some feedback on this new approach.